pkg/connector/group.go

```diff
@@ -250,112 +348,334 @@ func (g *groupBuilder) Grants(ctx context.Context, resource *v2.Resource, pToken
+				zap.String("principal_type", principal.Id.ResourceType),
+			)
+			return nil, fmt.Errorf("bitbucket(dc)-connector: user is already a member of the group")
+		}
 
-	groupPos := slices.IndexFunc(listGroup, func(c client.Members) bool {
-		return c.ID == userId
-	})
-	if groupPos != NF {
-		l.Warn(
-			"bitbucket(dc)-connector: user is already a member of the group",
-			zap.String("principal_id", principal.Id.String()),
-			zap.String("principal_type", principal.Id.ResourceType),
+		// Add user to the group
+		err = g.client.AddUserToGroups(ctx, groupName, userName)
+		if err != nil {
+			return nil, fmt.Errorf("bitbucket(dc)-connector: failed to add user to group: %w", err)
+		}
+
+		l.Warn("Membership has been created.",
+			zap.Int64("UserID", int64(userId)),
+			zap.String("User", userName),
+			zap.String("Group", groupName),
 		)
-		return nil, fmt.Errorf("bitbucket(dc)-connector: user is already a member of the group")
-	}
+	case resourceTypeRepository.Id:
+		groupName := entitlement.Resource.Id.Resource
+		_, permissions, err := ParseEntitlementID(entitlement.Id)
+		if err != nil {
+			return nil, err
+		}
 
-	// Add user to the group
-	err = g.client.AddUserToGroups(ctx, groupResourceId.Resource, principal.DisplayName)
-	if err != nil {
-		return nil, fmt.Errorf("bitbucket(dc)-connector: failed to add user to group: %w", err)
-	}
+		switch permissions[len(permissions)-1] {
+		case roleRepoWrite, roleRepoAdmin, roleRepoRead:
+			permission = permissions[len(permissions)-1]
+		default:
+			return nil, fmt.Errorf("bitbucket(dc)-connector: invalid permission type: %s", permissions[len(permissions)-1])
+		}
 
-	l.Warn("Membership has been created.",
-		zap.Int64("UserID", int64(userId)),
-		zap.String("User", principal.DisplayName),
-		zap.String("Group", groupResourceId.Resource),
-	)
+		repoId, err := strconv.Atoi(principal.Id.Resource)
+		if err != nil {
+			return nil, err
+		}
+
+		projectKey, repositorySlug, _, err = getRepositoryData(ctx, g, repoId)
+		if err != nil {
+			return nil, err
+		}
+
+		groupRepositoryPermissions, err := listGroupRepositoryPermissions(ctx, g.client, projectKey, repositorySlug)
+		err = checkStatusUnauthorizedError(ctx, err)
+		if err != nil {
+			return nil, err
+		}
+
+		groupRepositoryPermissionPos := slices.IndexFunc(groupRepositoryPermissions, func(c client.GroupsPermissions) bool {
+			return c.Group.Name == groupName
+		})
+		if groupRepositoryPermissionPos != NF {
+			l.Warn(
+				"bitbucket(dc)-connector: group already has this repository permission",
+				zap.String("principal_id", principal.Id.String()),
+				zap.String("principal_type", principal.Id.ResourceType),
+			)
+			return nil, fmt.Errorf("bitbucket(dc)-connector: group %s already has this repository permission", groupName)
+		}
+
+		err = g.client.UpdateGroupRepositoryPermission(ctx,
+			projectKey,
+			repositorySlug,
+			groupName,
+			permission,
+		)
+		err = checkStatusUnauthorizedError(ctx, err)
```

Comment:
If we get unauthorized in a grant or revoke, don't we want to bubble that error up so that it's clear the grant/revoke failed?

Reply:
Fixed